Add AOT annotations to System.Private.Xml and System.Data.Common

sbomer opened the pull request:
Contributes to #75480

sbomer committed:
Add AOT annotations to System.Private.Xml

sbomer committed:
Add more annotations

sbomer committed:
Remove DAM from TypeDesc, add more annotations

DAM on TypeDesc would require bubbling up RDC to parts of
TypeScope (AddPrimitive of byte[]) that was unnecessarily viral.

sbomer committed:
Update ref assemblies

sbomer committed:
Clean up System.Data.Common annotations

sbomer committed:
Annotate System.Data.Common

sbomer committed:
Merge remote-tracking branch 'origin/main' into aotSystemPrivateXml

eerhardt commented on src/libraries/System.Data.Common/src/System/Xml/XmlBoundElement.cs:
```diff
@@ -89,6 +90,7 @@ public override XmlNode? NextSibling
                 XmlNode? next = base.NextSibling;
                 if (next == null)
                 {
+
```
```suggestion
```

eerhardt commented on src/libraries/System.Private.Xml/src/System/Xml/Serialization/CodeGenerator.cs:
```diff
@@ -429,6 +429,7 @@ internal void NewArray(Type elementType, object len)
             _ilGen!.Emit(OpCodes.Newarr, elementType);
         }
 
+        [RequiresDynamicCode(XmlSerializer.AotSerializationWarning)]
```
I would think this whole class needs to be annotated with `[RequiresDynamicCode]`.

eerhardt commented on src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs:
```diff
@@ -42,6 +42,7 @@ public ReflectionXmlSerializationReader(XmlMapping mapping, XmlReader xmlReader,
         }
 
         [RequiresUnreferencedCode(XmlSerializer.TrimSerializationWarning)]
+        [RequiresDynamicCode(XmlSerializer.AotSerializationWarning)]
```
Would it be easier if we just moved the RUC and RDC attributes to the whole class?

eerhardt replied:
We might be able to remove the suppressions at the top of this file on line 16 and line 33.

eerhardt replied:
I think we would do the same thing to `ReflectionXmlSerializationWriter` - annotate the whole class as RDC, RUC. Then you shouldn't get any more warnings on `StringTypeDesc` and `QnameTypeDesc`

eerhardt replied:
Totally my opinion from looking at the code quick (I haven't done much development in this area). It is just easier to annotate and reason about when you slap the attributes on the whole class when the class's whole purpose is to do trim/AOT-incompatible stuff. In both these cases ("ReflectionReader/Writer" and "CodeGenerator") their whole purpose is to use reflection and IL Ref Emit.

eerhardt commented on src/libraries/System.Private.Xml/src/System/Xml/Serialization/Types.cs:
```diff
@@ -87,7 +87,6 @@ internal sealed class TypeDesc
         private TypeDesc? _nullableTypeDesc;
         private readonly TypeKind _kind;
         private readonly XmlSchemaType? _dataType;
-        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
```
Why remove these annotations? Typically we try to keep them if they "work" so if new code starts calling the methods it is annotated correctly.

MichalStrehovsky replied:
Would the newly added DAMT annotations be better for this (once I get that merged)? (I don't have an opinion about keeping the annotations or removing or updating it since I don't know what any of this is for - just bringing up one more option).

eerhardt commented on src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationReaderILGen.cs:
```diff
@@ -180,6 +180,7 @@ internal XmlSerializationReaderILGen(TypeScope[] scopes, string access, string c
         }
 
         [RequiresUnreferencedCode("calls WriteReflectionInit")]
+        [RequiresDynamicCode(XmlSerializer.AotSerializationWarning)]
```
It might be easier if we mark this whole class as RDC.

eerhardt commented on src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriter.cs:
```diff
@@ -83,6 +83,8 @@ protected ArrayList? Namespaces
                 return _namespaces?.NamespaceList;
             }
+
+            [RequiresDynamicCode(XmlSerializer.AotSerializationWarning)]
```
This can be suppressed right? We do it elsewhere in this PR.

eerhardt commented on src/libraries/System.Data.Common/src/System/Data/DataSet.cs:
```diff
@@ -306,6 +307,8 @@ protected DataSet(SerializationInfo info, StreamingContext context, bool Constru
 
         [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
             Justification = "Binary serialization is unsafe in general and is planned to be obsoleted. We do not want to mark interface or ctors of this class as unsafe as that would show many unnecessary warnings elsewhere.")]
+        [UnconditionalSuppressMessage("AotAnalysis", "IL3050:RequiresUnreferencedCode",
```
```suggestion
        [UnconditionalSuppressMessage("AotAnalysis", "IL3050:RequiresDynamicCode",
```

Repeated elsewhere.

sbomer committed:
PR feedback

- Fix whitespace
- Suppress warning for ToArray of known reference type
- RequiresUnreferencedCode -> RequiresDynamicCode

sbomer committed:
Annotate a few serialization-related types at the type level

sbomer committed:
Update ref assembly

sbomer committed:
Fix code analysis error

eerhardt reviewed (approved):
LGTM. Thanks for cleaning up the code where we could.
